amads/core/vectors_sets.py

```diff
@@ -0,0 +1,294 @@
+
+    >>> roundtrip = set_to_vector(resulting_set, max_index=6)
+    >>> roundtrip
+    (0, 3, 2, 1, 0, 0, 0)
+
+    >>> roundtrip == test_vector
+    True
+
+    """
+    return tuple(i for i in range(len(vector)) for _ in range(vector[i]))
+
+
+# Arithmetic operations: Addition/subtraction, multiplication,division
+
+
+def apply_constant(
+    set_or_vector: list[int] | tuple[int], constant: float, modulo: int | None = None
+) -> list | tuple:
+    """
+    Apply a constant value to a set or vector.
+
+    Parameters
+    ----------
+    set_or_vector: Any list or tuple representing a set or vector.
+    constant: The constant may be an int or a float and positive or negative.
+    modulo: Int or `None`. If int, return the values modulo this value (e.g., 12 for pitch class sets)
+
+    Returns
+    -------
+    list | tuple: The set or vector with constant applied.
+
+    Examples
+    --------
+
+    >>> start = [0, 1, 2, 3, 11]
+    >>> more = apply_constant(start, 1)
+    >>> more
+    [1, 2, 3, 4, 12]
+
+    >>> less = apply_constant(more, -1)
+    >>> less
+    [0, 1, 2, 3, 11]
+
+    >>> start == less
+    True
+
+    >>> more = apply_constant(start, 1, modulo=12)
+    >>> more
+    [1, 2, 3, 4, 0]
+    """
+    result = [x + constant for x in set_or_vector]
+    if modulo is not None:
+        result = [x % modulo for x in result]
+    return result
+
+
+def scalar_multiply(input_list: list, scale_factor: int = 2) -> list:
+    """
+    Multiply all values of a list.
+
+    Parameters
+    ----------
+    input_list: Any list.
+    scale_factor: The "scale factor" aka "multiplicative operand". Multiply all terms by this amount. Defaults to 2.
+
+    >>> scalar_multiply([0, 1, 2])
+    [0, 2, 4]
+
+    """
+    return [
+        x * scale_factor for x in input_list
+    ]  # TODO np would be better in cases like this
+
+
+# Transformations
+
+
+def rotate(vector: list[int] | tuple[int], steps: int | None = None) -> list:
+    """
+    Rotate a list by N steps.
+    This serves equivalently for
+    "phase shifting" of rhythm and
+    "transposition" of pitch.
+
+    Parameters
+    ----------
+    vector: Any list of any elements. We expect to work with a list integers representing a vector.
+    steps: how many steps to rotate.
+        Or, equivalently, the nth index of the input list becomes the 0th index of the new.
+        If unspecified, use the half cycle: int(<cycle lenth>/2).
+
+    Returns
+    -------
+    list | tuple: The input (list or tuple), rotated. Same length.
+
+    Examples
+    --------
+
+    >>> start = [0, 1, 2, 3]
+    >>> rotate(start, 1)
+    [1, 2, 3, 0]
+
+    >>> rotate(start, -1)
+    [3, 0, 1, 2]
+
+    >>> rotate([0, 1, 2, 3])  # note no steps specified
+    [2, 3, 0, 1]
+
+    """
+    if not steps:
+        steps = int(len(vector) / 2)
+
+    return vector[steps:] + vector[:steps]
+
+
+def mirror(vector: list | tuple, index_of_symmetry: int | None = None) -> list:
+    """
+    Reverse a list.
```

Comment:
Maybe 'reverse a vector'? I think it is good to be consistent about what we are calling our input object